src/cpu/flexcpu/inflight_inst.cc

```diff
@@ -0,0 +1,648 @@
+        // dependency was in-flight at the time that this instruction was
+        // issued.
+        return backingContext->readCCReg(reg_id.index());
+    }
+}
+
+void
+InflightInst::setCCRegOperand(const StaticInst* si, int dst_idx, CCReg val)
+{
+    const RegId& reg_id = si->destRegIdx(dst_idx);
+    assert(reg_id.isCCReg());
+
+    results[dst_idx].set(val, CCRegClass);
+    resultValid[dst_idx] = true;
+}
+
+MiscReg
+InflightInst::readMiscRegOperand(const StaticInst* si, int op_idx)
+{
+    const RegId& reg_id = si->srcRegIdx(op_idx);
+    assert(reg_id.isMiscReg());
+
+    const DataSource& source = sources[op_idx];
+    const shared_ptr<InflightInst> producer = source.producer.lock();
+
+    if (producer) {
+        // If the producing instruction is still in the buffer, grab the result
+        // assuming that it has been produced, and our index is in bounds.
+        assert(producer->isComplete());
+        assert(source.resultIdx >= 0
+            && source.resultIdx < producer->results.size()
+            && producer->resultValid[source.resultIdx]);
+        return producer->getResult(source.resultIdx).asMiscReg();
+    } else {
+        // Either the producing instruction has already been committed, or no
+        // dependency was in-flight at the time that this instruction was
+        // issued.
+        return backingContext->readMiscReg(reg_id.index());
+    }
+}
+
+void
+InflightInst::setMiscRegOperand(const StaticInst* si, int dst_idx,
+                                const MiscReg& val)
+{
+    const RegId& reg_id = si->destRegIdx(dst_idx);
+    assert(reg_id.isMiscReg());
+
+    results[dst_idx].set(val, MiscRegClass);
+    resultValid[dst_idx] = true;
+}
+
+MiscReg
+InflightInst::readMiscReg(int misc_reg)
+{
+    panic("readMiscReg() not implemented!");
+    return 0;
+    // TODO
+}
+
+void
+InflightInst::setMiscReg(int misc_reg, const MiscReg& val)
+{
+    panic("setMiscReg() not implemented!");
+    // TODO
+}
+
+PCState
+InflightInst::pcState() const
+{
+    return _pcState;
+}
+
+void
+InflightInst::pcState(const PCState& val)
+{
+    _pcState = val;
+}
+
+Fault
+InflightInst::readMem(Addr addr, uint8_t *data, unsigned int size,
+                Request::Flags flags)
+{
+    if (backingMemoryInterface) {
+        return backingMemoryInterface->readMem(shared_from_this(), addr, data,
+                                               size, flags);
+    } else {
+        panic("Attempted to readMem() without a memory interface!");
+        return NoFault;
+    }
+}
+
+Fault
+InflightInst::initiateMemRead(Addr addr, unsigned int size,
+                              Request::Flags flags)
+{
+    if (backingMemoryInterface) {
+        return backingMemoryInterface->initiateMemRead(shared_from_this(),
+                                                       addr, size, flags);
+    } else {
+        panic("Attempted to initiateMemRead() without a memory interface!");
```

Comment:
The constructor doesn't stop you from passing nullptr though, and I don't think it should stop you from passing nullptr, since it would be totally reasonable to construct a non-memref InflightInst and not to bother giving it an implementation of this interface.